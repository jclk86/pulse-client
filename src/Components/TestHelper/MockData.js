export const mockUser = {
  username: "Pop123",
  password: "Password123!",
  fullname: "Bob Smith",
  email: "BSmith@gmail.com"
};

export const mockArticle = {
  title: "test article",
  content: "test content",
  article_tag: "News",
  id: 1,
  date_created: "2019-10-14T21:49:20.570Z",
  author: {
    username: mockUser.username
  }
};

export const mockComment = {
  id: 1,
  content: "test comment content",
  date_created: "2019-10-14T21:49:20.570Z",
  article_id: 1,
  user: {
    username: mockArticle.author.username
  }
};
