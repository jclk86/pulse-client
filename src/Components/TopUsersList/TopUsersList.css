.top_users_list {
  padding: 0;
  width: 100%;
  margin: 0 auto;
}

.top_users_list_item {
  list-style: none;
  font-size: 16px;
  font-weight: bold;
  display: flex;
  margin-bottom: 20px;
  text-align: center;
  padding-top: 10px;
  padding-bottom: 5px;
}

.top_users_list_item:hover {
  background-color: rgba(109, 108, 108, 0.2);
}

.top_users_list_item_lights_off:hover {
  background-color: rgba(109, 108, 108, 0.2);
}

.container_top_user_username {
  width: 50%;
  text-align: left;
}

.top_users_link {
  text-decoration: none;
  color: black;
  word-wrap: break-word;
}

.top_users_link_lights_off {
  text-decoration: none;
  color: white;
  word-wrap: break-word;
}

.container_top_user_image {
  width: 25%;
}

.top_user_image {
  width: 20px;
  height: 20px;
  object-fit: cover;
  display: block;
  margin-left: auto;
  margin-right: 8px;
}

.container_top_user_chatbox {
  width: 25%;
}

.top_user_chatbox {
  width: 20px;
  margin-right: auto;
}
