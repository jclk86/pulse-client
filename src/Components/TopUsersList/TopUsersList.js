import React, { Component } from "react";
import { NavLink } from "react-router-dom";
import ArticleListContext from "../../Context/ArticleListContext";
import CommentApiService from "../../Services/comment-api-service";
import { sortByFrequencyAndRemoveDuplicates } from "../Utils/Utils";
import ChatBox from "../../images/chat.png";
import "./TopUsersList.css";

class TopUsersList extends Component {
  static contextType = ArticleListContext;

  componentDidMount() {
    CommentApiService.getAllComments().then(this.context.setComments);
  }

  getImagesForUsers = (comments, username) => {
    return comments.find(comment => {
      return comment.user.username === username ? comment.user.image_url : null;
    });
  };

  render() {
    const { comments, lightsOff } = this.context;
    const usersArray = comments.map(comment => comment.user.username);
    const topUsersList = sortByFrequencyAndRemoveDuplicates(usersArray);
    const topTenUsers = topUsersList.slice(0, 11);
    return (
      <ul className="top_users_list">
        {topTenUsers.map((user, i) => {
          const imageLinks = this.getImagesForUsers(comments, user);
          const userImage = imageLinks ? imageLinks.user.image_url : "";
          return (
            <li
              key={i}
              className={`top_users_list_item + ${
                lightsOff ? "" : "top_users_list_item_lights_off"
              }`}
            >
              <div className="container_top_user_image">
                <img
                  src={userImage}
                  alt="top user icon"
                  className="top_user_image"
                ></img>
              </div>
              <div className="container_top_user_username">
                {" "}
                <NavLink
                  to={`/profile/${user}`}
                  key={i}
                  className={`top_users_link + ${
                    lightsOff ? "" : "top_users_link_lights_off"
                  }`}
                >
                  {user}
                </NavLink>
              </div>
              <div className="container_top_user_chatbox">
                <img
                  src={ChatBox}
                  alt="top user icon"
                  className="top_user_chatbox"
                ></img>
              </div>
            </li>
          );
        })}
      </ul>
    );
  }
}

export default TopUsersList;
