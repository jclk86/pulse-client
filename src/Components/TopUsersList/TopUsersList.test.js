import React from "react";
import { BrowserRouter } from "react-router-dom";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import AuthApiService from "../../Services/auth-api-service";
import ArticleApiService from "../../Services/article-api-service";
import { mockUsers, mockArticles, mockComments } from "../TestHelper/MockData";
import TopUsersList from "./TopUsersList";

describe(`TopUsersList Component`, () => {
  beforeAll(() => {
    for (let i = 0; i < mockUsers.length; i++) {
      AuthApiService.postUser(mockUsers[i]).then(user => {
        AuthApiService.postLogin({
          username: user.username,
          password: user.password
        }).then(() => ArticleApiService.postArticle(mockArticles[i]));
      });
    }
  });

  it("renders TopUsersList component without issues", () => {
    const wrapper = mount(
      <BrowserRouter>
        <TopUsersList comments={mockComments}></TopUsersList>
      </BrowserRouter>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
