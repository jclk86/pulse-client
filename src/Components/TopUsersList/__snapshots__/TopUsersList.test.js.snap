// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopUsersList Component renders TopUsersList component without issues 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <TopUsersList
      comments={
        Array [
          Object {
            "article_id": 1,
            "content": "test comment content",
            "date_created": "2019-10-14T21:49:20.570Z",
            "id": 1,
            "user": Object {
              "id": 1,
              "username": "Pop123",
            },
          },
          Object {
            "article_id": 2,
            "content": "second test comment",
            "date_created": "2019-10-14T21:49:20.570Z",
            "id": 2,
            "user": Object {
              "id": 2,
              "username": "Pop123",
            },
          },
        ]
      }
    >
      <ul
        className="top_users_list"
      />
    </TopUsersList>
  </Router>
</BrowserRouter>
`;
